paderwasn/rir_simulation/rir_simulation.py

```diff
@@ -0,0 +1,525 @@
+    # according to target DRR for the currently drawn alpha, so adding
+    # additional reverberant parts causes an invalid solution
+    # if change_alpha is set after some iterations of rejection sampling,
+    # calculate alpha based on the early reflections of the RIR that will
+    # guarantee a solution during scaling to match the target DRR,
+    # but still matches the original range of alpha the best
+    if change_alpha:
+        # calculate DRR of ISM part
+        drr_im = (np.sum(rir_ism[lower_lim:upper_lim] ** 2)
+                  / np.sum(rir_ism[upper_lim:] ** 2))
+        # solve DRR formula to get alpha that would just create the
+        # DRR of the early reflections of the RIR, seen as upper bound for the
+        # maximum DRR to achieve, because only first reflections are simulated
+        # and amending of the stochastic part would overall lower the DRR
+        max_factor = (drr_im * np.pi * t60 * distance ** 2
+                      / np.prod(room_dimensions) * 100
+                      / directivity_attenuation ** 2)
+        if max_factor < alpha:
+            # for current alpha, target DRR would be more direct as DRR of
+            # early reflections, so re-draw alpha, that then creates a lower,
+            # but achievable target DRR and therefore a solution during
+            # later scaling can be found
+            if max_factor < lower_limit_alpha:
+                # use omnidirectional directivity factor (1) as lower limit
+                # because otherwise no solution can be found
+                alpha = np.random.uniform(1, max_factor)
+            else:
+                # sample alpha until valid solution found, that still lays
+                # within the original range, but preserves convergence
+                alpha = np.random.uniform(lower_limit_alpha, max_factor)
+                while alpha > max_factor:
+                    alpha = np.random.uniform(lower_limit_alpha, max_factor)
+
+    target_drr = calc_target_drr(room_dimensions, alpha, beta, t60,
+                                 directivity_attenuation, distance)
+    weighting_function =\
+        calc_window(delta, kappa, toa_direct, rir_len, sample_rate)
+
+    scaling = calc_scaling(rir_ism, raw_stochastic, weighting_function,
+                           target_drr, upper_lim, lower_lim)
+    if scaling is None:
+        return None, alpha
+    else:
+        rir = rir_ism + raw_stochastic * weighting_function * scaling
+        return rir, alpha
+
+
+def calc_rirs(
+        src_pos, sensor_pos, room_dimensions, t60, azimuth_directivity,
+        colatitude_directivity, src_directivity,
+        directivity_factor_range=(2.5, 5.5), kappa=1, beta=1., rir_len=16384,
+        max_order_ism=3, win_len_direct=40, sound_speed=343, sample_rate=16000,
+        alpha_diff=0.1, max_retries=10
+):
+    """
+    Room impulse response (RIR) simulation for microphone pairs as described in
```

Comment:
Room impulse response (RIR) simulation for microphone pairs, consisting of closely spaced microphones,  as described in